pkg/readiness/pruner/pruner_test.go

```diff
@@ -0,0 +1,270 @@
+package pruner
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	frameworksexternaldata "github.com/open-policy-agent/frameworks/constraint/pkg/externaldata"
+	configv1alpha1 "github.com/open-policy-agent/gatekeeper/v3/apis/config/v1alpha1"
+	syncsetv1alpha1 "github.com/open-policy-agent/gatekeeper/v3/apis/syncset/v1alpha1"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/cachemanager"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/cachemanager/aggregator"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/controller"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/controller/config/process"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/expansion"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/fakes"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/mutation"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/readiness"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/syncutil"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/watch"
+	testclient "github.com/open-policy-agent/gatekeeper/v3/test/clients"
+	"github.com/open-policy-agent/gatekeeper/v3/test/testutils"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	timeout = 10 * time.Second
+	tick    = 1 * time.Second
+)
+
+var cfg *rest.Config
+
+var (
+	syncsetGVK = syncsetv1alpha1.GroupVersion.WithKind("SyncSet")
+	configGVK  = configv1alpha1.GroupVersion.WithKind("Config")
+
+	configMapGVK = schema.GroupVersionKind{Version: "v1", Kind: "ConfigMap"}
+	podGVK       = schema.GroupVersionKind{Version: "v1", Kind: "Pod"}
+)
+
+func TestMain(m *testing.M) {
+	testutils.StartControlPlane(m, &cfg, 3)
+}
+
+type testOptions struct {
+	addConstrollers       bool
+	addExpectationsPruner bool
+	testLister            readiness.Lister
+}
+
+type testResources struct {
+	expectationsPruner *ExpectationsPruner
+	manager            manager.Manager
+	k8sClient          client.Client
+}
+
+func setupTest(ctx context.Context, t *testing.T, o testOptions) *testResources {
+	t.Helper()
+
+	mgr, wm := testutils.SetupManager(t, cfg)
+	c := testclient.NewRetryClient(mgr.GetClient())
+
+	var tracker *readiness.Tracker
+	var err error
+	if o.testLister != nil {
+		tracker = readiness.NewTracker(o.testLister, false, false, false)
+		require.NoError(t, mgr.Add(manager.RunnableFunc(func(ctx context.Context) error {
+			return tracker.Run(ctx)
+		})), "adding tracker to manager")
+	} else {
+		tracker, err = readiness.SetupTrackerNoReadyz(mgr, false, false, false)
```

Comment:
Let's just call readiness.NewTracker() like above, so it's clear how these two options behave differently.